src/openai.ts

```diff
@@ -76,66 +84,133 @@ export class McpxOpenAI {
+      logFinalMessage(messageIdx, messages)
+      return { response,  messages, index: messageIdx, done: true }
+    }
 
-      for (; messageIdx < messages.length; ++messageIdx) {
-        this.#logger.info({ exchange: messages[messageIdx] }, 'message')
+    for (; messageIdx < messages.length; ++messageIdx) {
+      this.#logger.info({ exchange: messages[messageIdx] }, 'message')
+    }
+
+    for (const tool of choice.message.tool_calls) {
+      if (tool.type !== 'function') {
+        return { response,  messages, index: messageIdx, done: false }
       }
 
-      for (const tool of choice.message.tool_calls) {
-        if (tool.type !== 'function') {
-          continue
-        }
-
-        try {
-          const abortcontroller = new AbortController()
-          const result = await this.#session.handleCallTool(
-            {
-              method: 'tools/call',
-              params: {
-                name: tool.function.name,
-                arguments: JSON.parse(tool.function.arguments),
-              },
+      try {
+        const abortcontroller = new AbortController()
+        const result = await this.#session.handleCallTool(
+          {
+            method: 'tools/call',
+            params: {
+              name: tool.function.name,
+              arguments: JSON.parse(tool.function.arguments),
             },
-            { signal: abortcontroller.signal }
-          )
-
-          messages.push({
-            role: 'tool',
-            content: JSON.stringify(result),
-            tool_call_id: tool.id,
-          })
-        } catch (err: any) {
-          messages.push({
-            role: 'tool',
-            content: err.toString(),
-            tool_call_id: tool.id,
-          })
-        }
+          },
+          { signal: abortcontroller.signal }
+        )
+
+        messages.push({
+          role: 'tool',
+          content: JSON.stringify(result),
+          tool_call_id: tool.id,
+        })
+      } catch (err: any) {
+        messages.push({
+          role: 'tool',
+          content: err.toString(),
+          tool_call_id: tool.id,
+        })
       }
-    } while (1)
-    for (; messageIdx < messages.length - 1; ++messageIdx) {
-      this.#logger.info({ exchange: messages[messageIdx] }, 'message')
     }
-    this.#logger.info({ lastMessage: messages[messageIdx] }, 'final message')
-    return response
+
+    return { response,  messages, index: messageIdx, done: false }
+  }
+}
+
+export class ToolSchemaError extends Error {
+  static parse(err: any): any {
+    console.log(JSON.stringify(err))
+    const error = err?.error
+    const code = error?.code
+    if (error?.type === 'invalid_request_error' && code === 'invalid_function_parameters') {
+      // e.g. tools[0].function.parameters
+      const regex = /tools\[(\d+)\]\.function\.parameters/;
+      const match = error.param?.match(regex);
+      if (match) {
+        const index = match ? parseInt(match[1]) : -1
```

Comment:
```suggestion
        const index = parseInt(match[1], 10) || -1
```

(since we already check `if (match)`)

Reply:
BAD LLM! BAD 🗞️